src/pipe/filter.pipe.ts

```diff
@@ -20,62 +20,77 @@ import { Pipe, PipeTransform } from '@angular/core';
+                return input.filter(this.filterByString(currentFilter, field));
+            case 'object':
+                // filter by object ignores 'field' if specified
+                return input.filter(this.filterByObject(currentFilter));
+            default:
+                // 'symbol' && 'undefined'
+                return input.filter(this.filterDefault(currentFilter, field));
         }
     }
-    private filterByStringAndField(filter, field) {
-        return value => {
-            return !filter || (value[field] && value[field].toLowerCase().indexOf(filter.toLowerCase()) !== -1);
-        };
+
+    private filterByNumber(filter, field?) {
+        return value =>
+            (value && !filter) || (typeof value === 'object' && field)
+                ? value[field] && typeof value[field] === 'number' && value[field] === filter
+                : typeof value === 'number' && value === filter;
     }
 
-    // adapted from https://github.com/VadimDez/ng2-filter-pipe
-    private filterByString(filter) {
-        return value => {
-            return !filter || value.toLowerCase().indexOf(filter.toLowerCase()) !== -1;
-        };
+    private filterByBoolean(filter, field?) {
+        return value =>
+            typeof value === 'object' && field
+                ? value[field] && typeof value[field] === 'boolean' && value[field] === filter
+                : typeof value === 'boolean' && value === filter;
     }
 
-    private filterDefault(filter) {
-        return value => {
-            return !filter || filter === value;
-        };
+    private filterByString(filter, field?) {
+        return value =>
+            (value && !filter) || (typeof value === 'object' && field)
+                ? value[field] && typeof value[field] === 'string' && value[field] === filter
+                : typeof value === 'string' && value.toLowerCase().indexOf(filter.toLowerCase()) !== -1;
+    }
+
+    private filterDefault(filter, field?) {
+        return value => ((value && !filter) || (typeof value === 'object' && field) ? value[field] && filter === value : filter === value);
     }
 
     private filterByObject(filter) {
         return value => {
             const keys = Object.keys(filter);
-            for (const key of keys) {
-                const type = typeof value[key];
-                let isMatching;
+            let isMatching = false;
 
-                if (type === 'string') {
-                    isMatching = this.filterByString(filter[key])(value[key]);
-                } else if (type === 'object') {
-                    isMatching = this.filterByObject(filter[key])(value[key]);
-                } else {
-                    isMatching = this.filterDefault(filter[key])(value[key]);
-                }
-
-                if (!isMatching) {
-                    return false;
+            // all fields defined in filter object must match
+            for (const key of keys) {
+                switch (typeof filter[key]) {
```

Comment:
instead of a switch I would use multiple if with early returns 
exemple:
```
if(typeof filter[key] === 'number') { 
  return this.filterByNumber(filter[key])(value[key])
}
```